### DISCUSSION TITLE
How to properly throw an error and exit with a specific exit code?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [jmcdo29/nest-commander#886] [Q&A] [answered]
For example, when a parse method fails, I want to output an error (using a logger) and exit with an exit code. I can exit using `process.exit(2)` in the parse method, but is it the best way?

Is there some lifecycle events (similart to those of [Nest](https://docs.nestjs.com/fundamentals/lifecycle-events))? Those need to be enabled using `app.enableShutdownHooks()` in `main.ts` file, however, there seems no way to do the same with `CommandFactory.run()`.

It might be a good idea to somehow abstract this and define the exit codes in Nest Commander and append them to the help of each command. I’d suggest to output them all in global help, but output only a subset of exit codes for (sub-) commands (output only those that are used by a command). However, if we define an exit cope in decorators, we need to reference them in the methods themselves, which might complicate the implementation a bit.

**Comment by jmcdo29** [comment]
[I have this mentioned in the docs](https://nest-commander.jaymcdoniel.dev/en/features/factory/#error-handling) that there's nothing specific that nest-commander does, it's all at the `commander` level and you can pass your own error handler through the `CommandFactory.run()`. method.

I'll think if there's a way to abstract an error handler out of this to make it easier to work with

**Reply by tukusejssirs** [reply]
Well, yeah, I could use `errorHandler`, but can I enable and use Nest shutdown hooks using `app.enableShutdownHooks()` and creating an `ExitAppService` service that impements `OnModuleDestroy`? One thing I am not sure how should I notify it of the desired exit code? Using some kind of cache storage (or even using Redis)? :thinking:

Anyway, I have just found that we can [displaying an error](https://www.npmjs.com/package/commander#display-error) in `commander`, e.g. `program.error('Custom processing has failed', { exitCode: 2, code: 'my.custom.error' })`. Is there a way to do this in Nest Commander too?

**Reply by jmcdo29** [reply, marked as answer]
Yeah, you can access the commander instance (`program` in the above line) via [`@InjectCommander()`](https://nest-commander.jaymcdoniel.dev/en/features/commander/#getting-the-commander-instance) and then use `this.commander.error('Error message', options)`

**Reply by jmcdo29** [reply]
As for the shutdown hooks, it's technically now possible to enable those using the `CommandFactory.createWithoutRunning()` that I mentioned in your other question today.

**Reply by tukusejssirs** [reply]
I don’t think that I should use shutdown hooks in this case, as I think `commander` does the error outputting and exit code usage quite good. Still I think it might be a nice addition to Nest Commander to collect all the error codes and list them in the help. For now, I need to do this manually.

Anyway, do you re-export `Command` from `commander`? I could not find it, therefore I need to export it under a different name directly from `commander`.

```
import {Command as cmd} from 'commander'

constructor(@InjectCommander() private readonly commander: cmd) {
  super()
}
```

**Reply by jmcdo29** [reply]
Currently, no, it is not exported from `nest-commander`. I should look into doing that though

**Reply by tukusejssirs** [reply]
Yeah, I think it might be a good idea, however, we can always fall back to importing it from `commander`. :man_shrugging:

**Reply by tukusejssirs** [reply]
Anyway, thanks for your help! :pray:
